kernel/include/kswap.h

```diff
@@ -68,11 +70,18 @@ static ALWAYS_INLINE unsigned int do_swap(unsigned int key,
 
 	z_check_stack_sentinel();
 
-	if (is_spinlock && lock != NULL) {
+	/* We always take the scheduler spinlock if we don't already
+	 * have it.  We "release" other spinlocks here.  But we never
+	 * drop the interrupt lock.
+	 */
+	if (is_spinlock && lock != NULL && lock != &sched_spinlock) {
 		k_spin_release(lock);
 	}
+	if (!is_spinlock || lock != &sched_spinlock) {
+		(void) k_spin_lock(&sched_spinlock);
+	}
```

Comment:
The above logic will (seemingly) misbehave if

```
is_spinlock=true;
lock=NULL;
```

It'll take the `sched_spinlock` but discard interrupt state, likely leaving us interrupt-disabled forever.

If this is expected to never come up, is there a reason for having a separate `is_spinlock` as opposed to just using `NULL` in this case?

If this is expected to come up, what are the desired semantics?

Reply:
This is subtle (I mean, it's the guts of the context switch, that's sort of a given).  But in fact no, the **outer** key state that we were passed as an argument is what gets restored.  That is, we're assuming details about the spinlock key implementation: it's the same value as returned by the underlying arch_irq_lock() primitive, which is recursive in this way.  Back in the uniprocessor days, irq_lock() was all we had, so that's the primitive architectures are expected to provide and SMP is built on top of that (and then had to provide a compatible implementation of irq_lock() that worked across CPUs, heh).

This would never be allowed in application code, but I'm OK with it here.

And the bit about "is_spinlock" is a holdover from the same era.  We used to pass an irq_lock() key to (what at the time was called) _Swap().  And the SMP version still supports that, so we actually have two versions of swap, one for spinlocks and one for traditional irq_locks.  So this function gets expanded twice, with is_spinlock used as a compile-time value to select which specific locking strategy we use.

(FWIW: I just checked, there's only two spots in the posix subsystem still using the old primitive.  So we should just port them to spinlocks and remove this.)